### PR TITLE
[MIR] Serialize MachineFrameInfo::isCalleeSavedInfoValid()

### PR DESCRIPTION
**Pull request by dtellenbach**
In case of functions without a stack frame no "stack" field is serialized into MIR which leads to isCalleeSavedInfoValid being false when reading a MIR file back in. To fix this we should serialize MachineFrameInfo::isCalleeSavedInfoValid() into MIR.

### CODE REVIEW

**Review by arsenm — COMMENTED** [review]
Needs a parsing test where this is true. Can you also add one that hit the issue you described? Can we just print the empty list in this case?

**Comment by francisvm** [comment]
> > Can we just print the empty list in this case?
> 
> That wouldn't help because the logic in MIRParser would still not find any callee-saved info and set it as not valid.

Would an empty stack list, `isCalleeSavedInfoValid: true` and `-run-pass none` test the right thing?

**Comment by dtellenbach** [comment]
> > > Can we just print the empty list in this case?
> > 
> > 
> > That wouldn't help because the logic in MIRParser would still not find any callee-saved info and set it as not valid.
> 
> Would an empty stack list, `isCalleeSavedInfoValid: true` and `-run-pass none` test the right thing?

Oh yes, testing is no issue. I understood the question as can't we omit `isCalleeSavedInfoValid` when just using an empty stack list.

**Comment by dtellenbach** [comment]
Added a test that should cover all combinations of having a stack or not and having valid callee-saved info or not.

File: llvm/test/CodeGen/MIR/Generic/calleesaveinfovalid.mir [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+  hasTailCall:     true
+  isCalleeSavedInfoValid: false
+stack: []
+machineFunctionInfo:
+  hasRedZone:      false
+body:             |
+  bb.0 (%ir-block.1):
+    liveins: $w0
+  
+    TCRETURNdi @f4, 0, csr_darwin_aarch64_aapcs, implicit $sp, implicit $w0
```
Comment (francisvm) [inline llvm/test/CodeGen/MIR/Generic/calleesaveinfovalid.mir:58]: You don't really need this. I think even the body is optional? (see llvm/test/CodeGen/MIR/Generic/llvmIRMissing.mir) If you really need something in the body you can use `$x0 = IMPLICIT_DEF`. I think the same applies to most of the other fields, and the entire LLVM IR.

Comment (dtellenbach) [reply llvm/test/CodeGen/MIR/Generic/calleesaveinfovalid.mir]: Oh right, since this uses `-run-pass=none` it doesn't need anything. Should be good now.

**Comment by dtellenbach** [comment]
Sure, the newly added test is failing because `llc` uses the x86 runner's target triple which doesn't know `$lr`. Moving to `MIR/AArch64` and requiring an arm64 triple.

File: llvm/test/CodeGen/MIR/AArch64/calleesavedinfovalid.mir [hunk 2]
```diff
@@ -0,0 +1,41 @@
+
+# RUN: llc -run-pass=none -mtriple arm64-apple-ios -o - %s | FileCheck %s
```
Comment (francisvm) [inline llvm/test/CodeGen/MIR/AArch64/calleesavedinfovalid.mir:2]: `-mtriple=aarch64--` to be consistent with the other CodeGen/MIR/AArch64 tests.

**Review by francisvm — APPROVED** [review]
LGTM

File: llvm/test/CodeGen/MIR/AArch64/calleesavedinfovalid.mir (hunk 2 above)
Comment (dtellenbach) [reply llvm/test/CodeGen/MIR/AArch64/calleesavedinfovalid.mir]: Done!

**Pull request merged**